src/Proto.Cluster/SingleNode/SingleNodeActivatorActor.cs

```diff
@@ -0,0 +1,215 @@
+
+
+    private readonly Cluster _cluster;
+    private readonly Dictionary<ClusterIdentity, PID> _actors = new();
+    private readonly HashSet<ClusterIdentity> _inFlightIdentityChecks = new();
+
+    public SingleNodeActivatorActor(Cluster cluster) => _cluster = cluster;
+
+    private Task OnStarted(IContext context)
+    {
+        var self = context.Self;
+        _cluster.System.EventStream.Subscribe<ActivationTerminated>(e => _cluster.System.Root.Send(self, e));
+        _cluster.System.EventStream.Subscribe<ActivationTerminating>(e => _cluster.System.Root.Send(self, e));
+
+        return Task.CompletedTask;
+    }
+
+    public Task ReceiveAsync(IContext context) =>
+        context.Message switch
+        {
+            Started                   => OnStarted(context),
+            Stopping                  => OnStopping(context),
+            ActivationRequest msg     => OnActivationRequest(msg, context),
+            ActivationTerminated msg  => OnActivationTerminated(msg),
+            ActivationTerminating msg => OnActivationTerminating(msg),
+            _                         => Task.CompletedTask
+        };
+
+    private async Task OnStopping(IContext context)
+    {
+        await StopActors(context);
+        _cluster.PidCache.RemoveByPredicate(kv => kv.Value.Address.Equals(context.System.Address, StringComparison.Ordinal));
+    }
+
+    private async Task StopActors(IContext context)
+    {
+        var stopping = new List<Task>();
+
+        var clusterIdentities = _actors.Keys.ToList();
+
+        foreach (var ci in clusterIdentities)
+        {
+            var pid = _actors[ci];
+            var stoppingTask = context.PoisonAsync(pid);
+            stopping.Add(stoppingTask);
+            _actors.Remove(ci);
+        }
+
+        //await graceful shutdown of all actors we no longer own
+        await Task.WhenAll(stopping);
+        Logger.LogInformation("[SingleNode] - Stopped {ActorCount} actors", clusterIdentities.Count);
+    }
+
+    private Task OnActivationTerminated(ActivationTerminated msg)
+    {
+        _cluster.PidCache.RemoveByVal(msg.ClusterIdentity, msg.Pid);
+
+        // we get this via broadcast to all nodes, remove if we have it, or ignore
+        if (Logger.IsEnabled(LogLevel.Trace))
+            Logger.LogTrace("[SingleNode] Terminated {Pid}", msg.Pid);
+
+        return Task.CompletedTask;
+    }
+
+    private Task OnActivationTerminating(ActivationTerminating msg)
+    {
+        // ActivationTerminating is sent to the local EventStream when a
+        // local cluster actor stops.
+
+        if (!_actors.ContainsKey(msg.ClusterIdentity))
+            return Task.CompletedTask;
+
+        if (Logger.IsEnabled(LogLevel.Trace))
+            Logger.LogTrace("[SingleNode] Terminating {Pid}", msg.Pid);
+
+        _actors.Remove(msg.ClusterIdentity);
+
+        // Broadcast ActivationTerminated to all nodes so that PidCaches gets
```

Comment:
Perhaps change the comment to reflect the fact that there is a single node? Also just send the event to the EventStream directly, skipping the broadcast (to make the code easier to reason about).

Reply:
Cleared up, thanks 😄